File: contracts/ts/email.ts
/**
 * IEmailService sends transactional emails triggered by user actions.
 * Do NOT use this for marketing or bulk email — use a dedicated platform instead.
 */
export interface IEmailService {
  /**
   * Sends a single transactional email.
   * Returns the provider-assigned message ID on success.
   */
  send(msg: EmailMessage): Promise<EmailResult>;

  /**
   * Sends multiple emails in a single API call.
   * Providers may impose limits on batch size — check provider docs.
   */
  sendBatch(msgs: EmailMessage[]): Promise<EmailResult[]>;
}

/** A single email to send. */
export interface EmailMessage {
  /** List of recipient email addresses. */
  to: string[];

  /** Sender address. Must be a verified domain/address in your provider. */
  from: string;

  /** Optional reply-to address. */
  replyTo?: string;

  /** Email subject line. */
  subject: string;

  /** Email content. Provide both html and text for best deliverability. */
  body: EmailBody;

  /** Additional email headers (e.g. List-Unsubscribe). */
  headers?: Record<string, string>;

  /** Provider-specific tags for analytics/filtering. */
  tags?: Record<string, string>;
}

/** Holds the content of an email in multiple formats. */
export interface EmailBody {
  /** HTML version of the email body. */
  html?: string;

  /** Plain-text version. Always provide this as a fallback. */
  text: string;
}

/** Result of a successful send call. */
export interface EmailResult {
  /** Provider-assigned message ID. Store this for delivery tracking. */
  messageId: string;
}

File: contracts/go/email.go
package contracts

import "context"

// EmailService sends transactional emails triggered by user actions.
// Do NOT use this for marketing or bulk email — use a dedicated marketing platform instead.
type EmailService interface {
	// Send sends a single transactional email.
	// Returns the provider-assigned message ID on success.
	Send(ctx context.Context, msg EmailMessage) (*EmailResult, error)

	// SendBatch sends multiple emails in a single API call.
	// Providers may impose limits on batch size — check provider docs.
	SendBatch(ctx context.Context, msgs []EmailMessage) ([]*EmailResult, error)
}

// EmailMessage is a single email to send.
type EmailMessage struct {
	// To is the list of recipient email addresses.
	To []string

	// From is the sender address. Must be a verified domain/address in your provider.
	From string

	// ReplyTo is the optional reply-to address.
	ReplyTo string

	// Subject is the email subject line.
	Subject string

	// Body holds the email content. Provide both HTML and Text for best deliverability.
	Body EmailBody

	// Headers holds additional email headers (e.g. List-Unsubscribe).
	Headers map[string]string

	// Tags hold provider-specific tags for analytics/filtering.
	Tags map[string]string
}

// EmailBody holds the content of an email in multiple formats.
type EmailBody struct {
	// HTML is the HTML version of the email body.
	HTML string

	// Text is the plain-text version. Always provide this as a fallback.
	Text string
}

// EmailResult is the result of a successful Send call.
type EmailResult struct {
	// MessageID is the provider-assigned message ID. Store this for delivery tracking.
	MessageID string
}
